docs/RFCS/20201020_sql_query_denylist.md

```diff
@@ -0,0 +1,183 @@
+- Feature Name: SQL Query Denylist
+- Status: draft
+- Start Date: 2020-10-20
+- Authors: angelapwen, otan
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: [51643](https://github.com/cockroachdb/cockroach/issues/51643)
+
+# Summary
+
+We propose a mechanism built into CockroachDB to prevent operator-defined SQL queries from being executed. This will immediately address the situations where clusters enter unresponsive states due to a query, or a query pattern, that has been identified. To bring the clusters back up as soon as possible and prevent them from going down again due to the same query, we need a query denylist on the database. This RFC details a file-managed SQL query denylist implementation that will allow database administrators to trigger query denial by uploading a YAML file comprising regular expression patterns to a customizable file location.  
+# Motivation
+
+It is important to support a file-managed denylist so that database administrators are able to quickly deny queries that have been confirmed to bring down clusters to the point of unresponsiveness. In this situation, SQL commands cannot be run from the console and cluster settings will not be able to be set. 
+
+Without this functionality, there may be applications endlessly retrying the “bad” queries that will continuously bring the cluster down when executed. There is currently no built-in way to stop execution of a query like this other than by preventing the query from being run in the first place, and it is time- and resource-intensive for developers to write their own. 
+
+As blocking a query pattern is invasive and potentially dangerous, it is also important to add a staging denylist for administrators to test whether the query pattern accurately matches the queries they meant to deny or not. 
+
+Having a query denylist will allow database administrators to mitigate these situations quickly and immediately without making any code changes. Additionally, this will give more flexibility and time for CockroachDB’s database and release engineers to work on database fixes relevant to the query. 
+
+# Guide-level explanation
+
+## New named concepts
+
+### Denylist
+The denylist we define for this RFC and plan to implement is a global, non-configurable SQL query denylist in CockroachDB. A location for this denylist will be configured via a CLI flag, and if a denylist is uploaded to that location, any following queries matching any pattern on the denylist file will be denied. The supported file format for the denylist is YAML; each query pattern on the denylist file must be specified via regular expression conventions. 
```

Comment:
"non-configurable" is confusing here, I assume you mean non-configurable through SQL

Reply:
Hm, yes you're right because evidently it can be configured through modifying the YAML file 😄 I'll think of better wording for this or remove the word altogether.